src/remote_write/src/pooled_parser.rs

```diff
@@ -0,0 +1,221 @@
+//! ```
+//!
+//! Wrong usage example:
+//!
+//! ```rust,ignore
+//! // WRONG: returned parsed data outlives the input.
+//! fn bar() -> PooledWriteRequest {
+//!     let raw_data = load_protobuf_data(); // Bytes
+//!     let parser = PooledParser::default();
+//!     parser.decode(raw_data).unwrap()
+//!     // The input `raw_data` is dropped here, but the returned `PooledWriteRequest` outlives it.
+//! }
+//! ```
+
+use bytes::Bytes;
+
+use crate::{
+    pb_reader::read_write_request,
+    pooled_types::{PooledWriteRequest, WriteRequestManager, POOL},
+    repeated_field::Clear,
+};
+
+#[derive(Debug, Clone)]
+pub struct PooledParser;
+
+impl PooledParser {
+    fn new() -> Self {
+        Self
+    }
+
+    /// Decode a [`PooledWriteRequest`] from the buffer and return it.
+    ///
+    /// This method will allocate a new [`PooledWriteRequest`] instance since it
+    /// is unable to use the object pool in sync functions.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use bytes::Bytes;
+    /// use remote_write::pooled_parser::PooledParser;
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// # let data = Bytes::new();
+    ///
+    /// let parser = PooledParser::default();
+    /// let request = parser.decode(data)?;
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn decode(&self, buf: Bytes) -> Result<PooledWriteRequest, String> {
+        // Cannot get a PooledWriteRequest instance from the pool in sync functions.
+        let mut request = PooledWriteRequest::default();
+        read_write_request(buf, &mut request).map_err(|e| e.to_string())?;
+        Ok(request)
+    }
+
+    /// Decode a [`PooledWriteRequest`] from the buffer and return a
+    /// [`deadpool::managed::Object`] wrapper.
+    ///
+    /// This method will reuse a [`PooledWriteRequest`] instance from the object
+    /// pool. After the returned object is dropped, it will be returned to the
+    /// pool.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use bytes::Bytes;
+    /// use remote_write::pooled_parser::PooledParser;
+    /// # use std::error::Error;
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), Box<dyn Error>> {
+    /// # let data = Bytes::new();
+    ///
+    /// let parser = PooledParser::default();
+    /// let request = parser.decode_async(data).await?;
+    ///
+    /// // Access the parsed data through the deadpool object wrapper.
+    /// println!("Parsed {} timeseries", request.timeseries.len());
+    /// for ts in &request.timeseries {
+    ///     for label in &ts.labels {
+    ///         let name = String::from_utf8_lossy(&label.name);
+    ///         let value = String::from_utf8_lossy(&label.value);
+    ///         println!("Label: {}={}", name, value);
+    ///     }
+    /// }
+    /// // Object automatically returned to the pool when dropped.
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn decode_async(
```

Comment:
We don't need this async functions, the encode/decode is a CPU-bound work, use async has little benefits.